### PR TITLE
persist: set a timeout on compaction requests

### PR DESCRIPTION
**Pull request by pH14**
I'll fill in more details on Monday, but it seems like we're seeing compaction requests occasionally start but never complete. This causes compaction for a shard to jam indefinitely, leading to a domino chain of greater problems, like `environmentd` crash looping. 

Inferring from what data points we do have, but lacking hard evidence, my best guess right now is some S3 calls got stuck indefinitely. We don't currently set any client timeouts on our calls, so it seems possible that a call could be made to an unreachable destination and never complete. We could add timeouts directly to the S3 client, but https://github.com/awslabs/smithy-rs/pull/1740 is upcoming and completely overhauls the timeout structure, and seems much improved over what we could do on today's SDK version.

This PR puts a hard timeout on compaction, so that requests over 10 minutes will be dropped. This should prevent us from getting stuck indefinitely on a long request, at the risk of dropping/spinning on requests that really do need 10 minutes. In practice, the mean compaction time is under a second, but, just want to highlight the potential risk here. 

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

File: src/persist-client/src/internal/compact.rs [hunk 1]
```diff
@@ -118,29 +119,35 @@ where
...
-                            || "persist::compact::consolidate",
-                            Self::compact(
-                                cfg.clone(),
-                                Arc::clone(&blob),
-                                Arc::clone(&metrics),
-                                Arc::clone(&cpu_heavy_runtime),
-                                req,
-                                writer_id,
+                    let res = tokio::time::timeout(
+                        Duration::from_secs(60 * 10),
```
Comment (danhhz) [inline src/persist-client/src/internal/compact.rs:123]: is it crazy to do something like `max(60s, 1s per MiB)`? that would hopefully be something we could keep forever

either way, let's put any constants here as things on PersistConfig (i'm basically always going to say this when I see a constant in persist code :)

File: src/persist-client/src/internal/compact.rs [hunk 2]
```diff
@@ -118,29 +119,35 @@ where
...
-                        .map_err(|err| anyhow!(err));
-
-                    metrics
-                        .compaction
-                        .seconds
-                        .inc_by(start.elapsed().as_secs_f64());
+                            .map_err(|e| anyhow!(e)),
+                    )
+                    .await
+                    .map_err(|e| anyhow!(e));
```
Comment (danhhz) [inline src/persist-client/src/internal/compact.rs:140]: we should also inc a counter on timeout

**Review by danhhz — APPROVED** [review]
LGTM modulo a couple commits

File: src/persist-client/src/internal/compact.rs (hunk 1 above)
Comment (pH14) [reply src/persist-client/src/internal/compact.rs:123]: I like that! adding both the `max` and moving to cfg

**Pull request merged**